framework/src/org/checkerframework/framework/type/DefaultInferredTypesApplier.java

```diff
@@ -15,179 +19,111 @@
+            final Set<AnnotationMirror> inferredSet,
+            final TypeMirror inferredTypeMirror) {
+        if (inferredSet == null) {
+            return;
         }
+        for (final AnnotationMirror top : hierarchy.getTopAnnotations()) {
+            AnnotationMirror inferred = hierarchy.findAnnotationInHierarchy(inferredSet, top);
 
-        @Override
-        protected String defaultErrorMessage(
-                AnnotatedTypeMirror type, AnnotatedTypeMirror inferred, AnnotationMirror top) {
-            return "applyInferredToBoundedType: Unexpected AnnotatedTypeMirror combo:\n"
-                    + "type="
-                    + type
-                    + "\n"
-                    + "inferred="
-                    + inferred
-                    + "\n"
-                    + "top="
-                    + top
-                    + "\n";
+            apply(type, inferred, inferredTypeMirror, top);
         }
+    }
 
-        @Override
-        protected Void defaultAction(
-                AnnotatedTypeMirror type, AnnotatedTypeMirror inferred, AnnotationMirror top) {
-            final AnnotationMirror inferredAnnotation;
-            if (QualifierHierarchy.canHaveEmptyAnnotationSet(type)) {
-                inferredAnnotation = inferred.getAnnotationInHierarchy(top);
-            } else {
-                inferredAnnotation = inferred.getEffectiveAnnotationInHierarchy(top);
-            }
-            if (inferredAnnotation == null) {
-                // We inferred "no annotation" for this hierarchy.
-                type.removeAnnotationInHierarchy(top);
-            } else {
-                // We inferred an annotation.
-                AnnotationMirror present = type.getAnnotationInHierarchy(top);
-                if (present != null) {
-                    if (omitSubtypingCheck
-                            || qualifierHierarchy.isSubtype(inferredAnnotation, present)) {
-                        type.replaceAnnotation(inferredAnnotation);
-
-                    } else {
-                        // TODO: UNCOMMENT AND FIX UNARIES AND OTHER CASES THAT LEAD TO THIS ISSUE
-                        //                        ErrorReporter.errorAbort(
-                        //                            "Inferred type is above present type:\n"
-                        //                          + "inferredAnnotation=" + inferredAnnotation + "\n"
-                        //                          + "present=" + present
-                        //                        );
-                    }
-                } else {
-                    type.addAnnotation(inferredAnnotation);
-                }
+    private void apply(
+            AnnotatedTypeMirror type,
+            AnnotationMirror inferred,
+            TypeMirror inferredTypeMirror,
+            AnnotationMirror top) {
+
+        AnnotationMirror primary = type.getAnnotationInHierarchy(top);
+        if (inferred == null && primary == null) {
```

Comment:
I would find these much easier to read if the shared conditions are checked once, instead of repeatedly with additional conditions.